### ISSUE TITLE
Parallel Processing

### ISSUE THREAD

**Issue by tebeka** [tebeka/fastavro#6]
Ken said:

> Hey, I'm using your FastAvro module, and it's working great.   I'm processing network security data with it - about 500 million records a day right now, and that might grow to 5-20+ billion a day in a couple of years.
> 
> At that volume I need to squeeze a bit more performance out of my avro processing.  Do you have any thoughts you can share about the feasibility of using threads, multiprocessing or futures to support read parallelism at the block level?

**Comment by tebeka** [comment]
Probably some kind of [pmap](https://clojuredocs.org/clojure.core/pmap) or returning a iterable where multiple threads contribute to it.

**Comment by mjwillson** [comment]
Before worrying about parallel processing there's a lot more mileage you could get out of Cython.

From what I can see, right now Cython is only being used to compile the plain python implementations (reader.py / writer.py), which generates C code that still needs to do a lot of the same type-checking and busywork that the interpreter would. For me this results in code that's still very slow, especially at decoding things like arrays of ints.

Creating reader.pyx and writer.pyx and adding cython type annotations to them, using cdef rather than def for internal utility functions etc, should give a decent extra boost.

**Comment by wesm** [comment]
Is there some resistance to going ahead and doing that? I would be interested in starting to wrap the Avro-C API in Cython as well

**Comment by mjwillson** [comment]
No resistance that I know of. I would do it if I had time -- I'm afraid in my case I switched to a hand-coded C implementation wrapped in Cython for my small-ish special-purpose schema.

If you do go down the Cython path I'd probably still avoid the Avro C API and just directly decode the avro into python objects in C/Cython code. The original slow avro-python library takes the approach of wrapping the avro C API and the C API just adds overhead (lots of wrapper structs and its own reference counting IIRC) and gets in the way. Avro is actually quite a simple format to encode/decode directly, as you can see from the fact reader.py is only 250-ish SLOC of python.

**Comment by wesm** [comment]
There's talk of PyPy in the setup.py, but does anyone use this module with pypy? Any serious production use cases are going to be CPython in 99% of cases

**Comment by tebeka** [comment]
I agree that before going parallel we need to try to get out as much as we can from Cython. I'm by no means expert in Cython. 

A quick google search shows that Cython has some [pypy support](http://docs.cython.org/src/userguide/pypy.html). If it's not too much trouble I'd like to keep pypy support - however I agree with @wesm that's we can probably drop it. Python 3 support should stay though.

**Comment by wesm** [comment]
I can undertake a Cython "port" / consolidation sometime in the next month or two. I have a lot of experience with it and we'll be able to squeeze out a lot better performance by doing so

**Comment by tebeka** [comment]
@wesm That would be awesome - got for it! Thanks.

**Comment by rouge8** [comment]
As a counterexample, we use fastavro almost exclusively with PyPy.

**Comment by wesm** [comment]
PyPy support shouldn't stop further optimizations in Cython for CPython. In the worst case, we can maintain a pure Python module and an optimized Cython module.

**Comment by mohitapplift** [comment]
I am using fastavro for processing live log records. The frequency at which I am ingesting log records is around 1000 records/second. The problem I am facing is high CPU utilization. CPU utilisation is always ranging around 100%. Any suggestion is most welcome.

**Comment by tebeka** [comment]
@mohitapplift Thanks for reporting. However I feel this is not related to the current issue. Can you either provide more information on how is this related or open a new one?

**Comment by mohitapplift** [comment]
I am sorry. My issue is not related with current issue. I am opening a new one.

**Comment by barrywhart** [comment]
About Cython, I added a more Cython-native reader and writer implementation recently. Interestingly, the writer got about 2x faster while the reader only got about 25% faster. My guess is that the reader is CPU allocating all the small Python objects that make up Avro data converted back to Python. (I don't have any empirical evidence of this, however. [Cachegrind](http://valgrind.org/docs/manual/cg-manual.html) analysis, anyone?

About parallel processing, I'm curious what sort of processing you're doing on these files, and whether Python is the best tool for that processing. At MallChimp, we tend to use Python to create large Avro files which we then analyze in Hive or Spark. Both those tools have built-in support for splitting Avro files and processing them in parallel. On a mid-sized cluster, we can process 10 billion records in just a few hours. I would start out looking there. While I love Python, the Hadoop and JVM platforms tend to have better support for big data processing.

**Comment by scottbelden** [comment]
Cython performance should be better now than it was when this was opened and it's unclear what the bottleneck is. If things feel like they should be faster, we should profile and see where speed ups can be made. However, without any additional information, I feel like there's nothing that can be done here.

If there is a performance issue please open a new issue so that we can tackle that. Thanks.

**Comment by rmax** [comment]
FYI, spark-like parallel processing can be done with [dask](http://dask.pydata.org/en/latest/) and [dask-avro](https://github.com/rmax/dask-avro#dask-avro) which uses `fastavro` under the hood.
